internal/providers/util/retry.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Red Hat
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"context"
+	"time"
+)
+
+// RetryWithBackoff calls function f with an increasing exponential
+// delay on initialBackoff (up to maxBackoff) until timeout is reached
+// or it succeeds. If timeout is 0 then the f will be retried until it
+// succeeds. If f does not succeed after maxBackoff is achieved, the
+// error from the last call is returned.
+func RetryWithBackoff(initialBackoff, maxBackoff, timeout time.Duration, f func() error) error {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	if timeout != 0 {
+		cancelFn()
+		ctx, cancelFn = context.WithTimeout(context.Background(), timeout)
+	}
+	defer cancelFn()
+
+	var err error
+	duration := initialBackoff
+	for {
+		err = f()
+		if err == nil {
+			return nil
+		}
+
+		duration = duration * 2
```

Comment:
The initial backoff is never used.

Reply:
Ah yeah, I can move it to after the `select`.

Worth noting this code is only slightly modified from the code in [`httpReaderWithHeader`](https://github.com/coreos/ignition/blob/master/internal/resource/http.go#L294-L320) (which is intentionally not using this function as it wants to do additional logging using a loop counter), which has the same problem.

Reply:
Oof.  Probably worth updating that code also, and changing `initialBackoff` accordingly so we don't have a gratuitous behavior change.

Reply:
Done in latest patch set.